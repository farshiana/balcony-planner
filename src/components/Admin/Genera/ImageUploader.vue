<template>
    <div class="image-uploader">
        <template v-if="file">
            <div ref="croppie" />
        </template>
        <template v-else>
            <div
                class="upload-area d-flex align-center justify-center clickable"
                @click="$refs.imageInput.click()"
            >
                <span class="text-uppercase text-medium">{{ $t('shared.upload') }}</span>
            </div>
            <input
                ref="imageInput"
                type="file"
                accept="image/png"
                style="display: none"
                @change="onImageChange"
            >
        </template>
    </div>
</template>

<script>
import Croppie from 'croppie';
import 'croppie/croppie.css';

export default {
    data() {
        return {
            file: null,
            croppie: null,
        };
    },
    methods: {
        onImageChange(event) {
            [this.file] = event.target.files;

            const reader = new FileReader();
            reader.onload = (loadEvent) => {
                this.croppie = new Croppie(this.$refs.croppie, { // eslint-disable-line no-new
                    viewport: { width: 256, height: 256 },
                    url: loadEvent.target.result,
                });
            };

            reader.readAsDataURL(event.target.files[0]);
        },
        getResult() {
            return this.croppie.result({ type: 'base64' });
        },
    },
};
</script>

<style lang="scss" scoped>
.image-uploader {
    width: 100%;
    height: 400px;
    .upload-area {
        height: 100%;
        border: thin dashed #CFD8DC;
        border-radius: 4px;
        &:hover {
            background: #ECEFF1;
        }
    }
}
</style>
