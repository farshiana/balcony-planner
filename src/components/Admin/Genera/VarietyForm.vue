<template>
    <v-dialog v-model="dialog" max-width="600px">
        <v-card>
            <v-card-title>
                <span class="headline">
                    {{ variety.id ? $t('admin.varieties.editVariety') : $t('admin.varieties.newVariety') }}
                </span>
            </v-card-title>
            <v-form lazy-validation @submit.prevent="onSubmit">
                <v-card-text>
                    <v-container>
                        <v-row>
                            <v-col cols="12" sm="6" md="6">
                                <v-autocomplete
                                    v-model="variety.genusId"
                                    :items="genera"
                                    :error-messages="genusErrors"
                                    item-value="id"
                                    item-text="name"
                                    :label="$t('shared.genus')"
                                    clearable
                                    required
                                    @blur="$v.variety.genusId.$touch()"
                                />
                            </v-col>
                            <v-col cols="12" sm="6" md="6">
                                <v-text-field
                                    v-model="variety.name"
                                    :error-messages="nameErrors"
                                    :label="$t('admin.varieties.name')"
                                    required
                                    @blur="$v.variety.name.$touch()"
                                />
                            </v-col>
                            <v-col cols="12" sm="6" md="6">
                                <v-select
                                    v-model="variety.exposure"
                                    :error-messages="exposureErrors"
                                    :items="exposures"
                                    :label="$t('shared.exposure')"
                                    required
                                    @blur="$v.variety.exposure.$touch()"
                                />
                            </v-col>
                            <v-col cols="12" sm="6" md="6">
                                <v-select
                                    v-model="variety.watering"
                                    :error-messages="wateringErrors"
                                    :items="waterings"
                                    :label="$t('shared.watering')"
                                    required
                                    @blur="$v.variety.watering.$touch()"
                                />
                            </v-col>
                            <v-col cols="12" class="overflow-auto">
                                <p>{{ $t('shared.seed') }}</p>
                                <v-btn-toggle
                                    v-model="variety.seed"
                                    dense
                                    multiple
                                    rounded
                                    mandatory
                                >
                                    <v-btn
                                        v-for="(month, index) in shortMonths"
                                        :key="index"
                                        active-class="brown white--text"
                                        x-small
                                    >
                                        {{ month }}
                                    </v-btn>
                                </v-btn-toggle>
                            </v-col>
                            <v-col cols="12" class="overflow-auto">
                                <p>{{ $t('shared.plant') }}</p>
                                <v-btn-toggle
                                    v-model="variety.plant"
                                    dense
                                    multiple
                                    rounded
                                    mandatory
                                >
                                    <v-btn
                                        v-for="(month, index) in shortMonths"
                                        :key="index"
                                        active-class="orange white--text"
                                        x-small
                                    >
                                        {{ month }}
                                    </v-btn>
                                </v-btn-toggle>
                            </v-col>
                            <v-col cols="12" class="overflow-auto">
                                <p>{{ $t('shared.harvest') }}</p>
                                <v-btn-toggle
                                    v-model="variety.harvest"
                                    dense
                                    multiple
                                    rounded
                                    mandatory
                                >
                                    <v-btn
                                        v-for="(month, index) in shortMonths"
                                        :key="index"
                                        active-class="green white--text"
                                        x-small
                                    >
                                        {{ month }}
                                    </v-btn>
                                </v-btn-toggle>
                            </v-col>
                        </v-row>
                    </v-container>
                </v-card-text>
                <v-card-actions>
                    <v-spacer />
                    <v-btn color="info" text @click="dialog = false;">{{ $t('shared.cancel') }}</v-btn>
                    <v-btn color="primary" text type="submit" :loading="saving">{{ $t('shared.save') }}</v-btn>
                </v-card-actions>
            </v-form>
        </v-card>
    </v-dialog>
</template>

<script>
import { mapState, mapGetters, mapActions } from 'vuex';
import { validationMixin } from 'vuelidate';
import { required, maxLength } from 'vuelidate/lib/validators';
import { exposures, waterings } from '@/constants';
import { shortMonths } from '@/utils';

export default {
    mixins: [validationMixin],
    validations: {
        variety: {
            name: { required, maxLength: maxLength(30) },
            genusId: { required },
            exposure: { required },
            watering: { required },
        },
    },
    props: {
        visible: {
            type: Boolean,
            required: true,
        },
        variety: {
            type: Object,
            required: true,
        },
    },
    data() {
        return {
            shortMonths,
            exposures,
            waterings,
            saving: false,
        };
    },
    computed: {
        ...mapState('genera', ['genera']),
        ...mapGetters('genera', ['getGenusById']),

        dialog: {
            get() {
                return this.visible;
            },
            set(visible) {
                this.$emit('toggle', visible);
            },
        },
        genusErrors() {
            const errors = [];
            if (!this.$v.variety.genusId.$dirty) return errors;
            if (!this.$v.variety.genusId.required) errors.push(this.$t('genusRequired'));
            return errors;
        },
        nameErrors() {
            const errors = [];
            if (!this.$v.variety.name.$dirty) return errors;
            if (!this.$v.variety.name.maxLength) errors.push(this.$t('nameMaxLength'));
            if (!this.$v.variety.name.required) errors.push(this.$t('nameRequired'));
            return errors;
        },
        exposureErrors() {
            const errors = [];
            if (!this.$v.variety.exposure.$dirty) return errors;
            if (!this.$v.variety.exposure.required) errors.push(this.$t('exposureRequired'));
            return errors;
        },
        wateringErrors() {
            const errors = [];
            if (!this.$v.variety.watering.$dirty) return errors;
            if (!this.$v.variety.watering.required) errors.push(this.$t('wateringRequired'));
            return errors;
        },
    },
    methods: {
        ...mapActions('genera', ['addVariety', 'updateVariety']),

        async onSubmit() {
            this.$v.variety.$touch();
            if (this.$v.variety.$invalid || this.saving) return;

            this.saving = true;
            if (this.variety.id) {
                await this.updateVariety(this.variety);
            } else {
                await this.addVariety(this.variety);
            }
            this.saving = false;

            this.dialog = false;
            this.$v.variety.$reset();
        },
    },
};
</script>
