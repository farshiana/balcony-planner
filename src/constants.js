// CATEGORIES
export const CATEGORY_FRUITS = 'fruits';
export const CATEGORY_HERBS = 'herbs';
export const CATEGORY_VEGETABLES = 'vegetables';
export const CATEGORIES = [
    CATEGORY_FRUITS,
    CATEGORY_HERBS,
    CATEGORY_VEGETABLES,
];

// EXPOSURES
export const EXPOSURE_SHADE = 'shade';
export const EXPOSURE_SUN = 'sun';
export const EXPOSURE_PARTIAL = 'partial';
export const EXPOSURES = [
    EXPOSURE_SHADE,
    EXPOSURE_SUN,
    EXPOSURE_PARTIAL,
];

// WATERINGS
export const WATERING_LOW = 'low';
export const WATERING_MEDIUM = 'medium';
export const WATERING_HIGH = 'high';
export const WATERINGS = [
    WATERING_LOW,
    WATERING_MEDIUM,
    WATERING_HIGH,
];
