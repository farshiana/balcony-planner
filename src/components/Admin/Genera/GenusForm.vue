<template>
    <v-dialog v-model="dialog" max-width="600px">
        <v-card>
            <v-card-title>
                <span class="headline">
                    {{ genus.id ? $t('admin.genera.editGenus') : $t('admin.genera.newGenus') }}
                </span>
            </v-card-title>
            <v-form lazy-validation @submit.prevent="onSubmit">
                <v-card-text>
                    <v-container>
                        <v-row>
                            <v-col cols="12" sm="6" md="6">
                                <v-img v-if="genus.imageUrl" src="https://picsum.photos/id/11/500/300" />
                                <image-uploader v-else />
                            </v-col>
                            <v-col cols="12" sm="6" md="6">
                                <v-select
                                    v-model="genus.category"
                                    :items="categories"
                                    :error-messages="categoryErrors"
                                    :label="$t('admin.genera.category')"
                                    clearable
                                    required
                                    @blur="$v.genus.category.$touch()"
                                />
                                <v-text-field
                                    v-model="genus.name"
                                    :error-messages="nameErrors"
                                    :label="$t('shared.name')"
                                    required
                                    @blur="$v.genus.name.$touch()"
                                />
                            </v-col>
                        </v-row>
                    </v-container>
                </v-card-text>
                <v-card-actions>
                    <v-spacer />
                    <v-btn color="info" text @click="dialog = false;">{{ $t('shared.cancel') }}</v-btn>
                    <v-btn color="primary" text type="submit" :loading="saving">{{ $t('shared.save') }}</v-btn>
                </v-card-actions>
            </v-form>
        </v-card>
    </v-dialog>
</template>

<script>
import { mapActions } from 'vuex';
import { validationMixin } from 'vuelidate';
import { required, maxLength } from 'vuelidate/lib/validators';
import {
    CATEGORY_FRUITS,
    CATEGORY_HERBS,
    CATEGORY_VEGETABLES,
    API_URL,
} from '@/constants';
import ImageUploader from './ImageUploader.vue';

export default {
    components: {
        ImageUploader,
    },
    mixins: [validationMixin],
    validations: {
        genus: {
            category: { required },
            name: { required, maxLength: maxLength(30) },
            imageUrl: { required },
        },
    },
    props: {
        visible: {
            type: Boolean,
            required: true,
        },
        genus: {
            type: Object,
            required: true,
        },
    },
    data() {
        return {
            API_URL,
            saving: false,
            categories: [
                { value: CATEGORY_FRUITS, text: this.$t('shared.fruits') },
                { value: CATEGORY_HERBS, text: this.$t('shared.herbs') },
                { value: CATEGORY_VEGETABLES, text: this.$t('shared.vegetables') },
            ],
        };
    },
    computed: {
        dialog: {
            get() {
                return this.visible;
            },
            set(visible) {
                this.$emit('toggle', visible);
            },
        },
        categoryErrors() {
            const errors = [];
            if (!this.$v.genus.category.$dirty) return errors;
            if (!this.$v.genus.category.required) errors.push(this.$t('admin.genera.categoryRequired'));
            return errors;
        },
        nameErrors() {
            const errors = [];
            if (!this.$v.genus.name.$dirty) return errors;
            if (!this.$v.genus.name.maxLength) errors.push(this.$t('shared.nameMaxLength'));
            if (!this.$v.genus.name.required) errors.push(this.$t('shared.nameRequired'));
            return errors;
        },
    },
    methods: {
        ...mapActions('genera', ['addGenus', 'updateGenus']),

        async onSubmit() {
            this.$v.genus.$touch();
            if (this.$v.genus.$invalid || this.saving) return;

            this.saving = true;
            if (this.genus.id) {
                await this.updateGenus(this.genus);
            } else {
                await this.addGenus(this.genus);
            }
            this.saving = false;

            this.dialog = false;
            this.$v.genus.$reset();
        },
    },
};
</script>
