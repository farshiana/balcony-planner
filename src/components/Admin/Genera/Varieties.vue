<template>
    <v-card class="varieties ma-3">
        <v-card-title>
            {{ $t('shared.varieties') }}
            <v-spacer />
            <v-text-field
                v-model="search"
                append-icon="mdi-magnify"
                :label="$t('shared.searchPlaceholder')"
                single-line
                hide-details
            />
        </v-card-title>
        <v-data-table
            :headers="headers"
            :items="genus.varieties"
            :items-per-page="12"
            :loading="loading"
            dense
            :search="search"
        >
            <template  v-slot:top>
                <v-row fill-height class="mr-1">
                    <v-spacer />
                    <v-btn class="mt-1" text small color="accent" @click="onAdd">
                        {{ $t('admin.varieties.addVariety') }}
                    </v-btn>
                </v-row>
            </template>
            <template v-slot:[`item.name`]="{ item }">
                <div class="d-flex align-center">
                    <div class="mr-1"><v-img :src="item.imageUrl" width="12px" height="12px" /></div>
                    <span class="text-capitalize-first">{{ item.name }}</span>
                </div>
            </template>
            <template v-slot:[`item.exposure`]="{ item }">
                <div class="d-flex align-center">
                    <exposure :exposure="item.exposure" />
                </div>
            </template>
            <template v-slot:[`item.watering`]="{ item }">
                <div class="d-flex align-center">
                    <watering :watering="item.watering" />
                </div>
            </template>
            <template v-slot:[`header.timeline`]="">
                <div class="d-flex flex-row justify-space-around">
                    <div
                        v-for="month in shortMonths"
                        :key="month"
                    >
                        {{ month }}
                    </div>
                </div>
            </template>
            <template v-slot:[`item.timeline`]="{ item }">
                <timeline
                    :seed="item.seed"
                    :plant="item.plant"
                    :harvest="item.harvest"
                />
            </template>
            <template v-slot:[`item.actions`]="{ item }">
                <v-btn
                    icon
                    small
                    class="mr-2"
                    @click.stop="onEdit(item)"
                >
                    <v-icon small>mdi-pencil</v-icon>
                </v-btn>
                <v-btn
                    icon
                    small
                    @click.stop="onDelete(item)"
                >
                    <v-icon small>mdi-delete</v-icon>
                </v-btn>
            </template>
        </v-data-table>
        <variety-form
            v-if="dialog"
            :visible="dialog"
            :variety="variety"
            @toggle="(visible) => { dialog = visible; }"
        />
        <delete-dialog
            :visible.sync="deleteDialog"
            :title="$t('admin.varieties.deleteVariety')"
            :name="variety.name"
            :deleting="deleting"
            @delete="onConfirmDelete"
        />
    </v-card>
</template>

<script>
import Vue from 'vue';
import { mapActions } from 'vuex';
import { shortMonths } from '@/constants';
import Timeline from '@/components/Timeline.vue';
import Exposure from '@/components/Exposure.vue';
import Watering from '@/components/Watering.vue';
import DeleteDialog from '@/components/DeleteDialog.vue';
import VarietyForm from './VarietyForm.vue';

export default {
    components: {
        Timeline,
        VarietyForm,
        Exposure,
        Watering,
        DeleteDialog,
    },
    props: {
        genus: {
            type: Object,
            required: true,
        },
    },
    data() {
        return {
            search: '',
            shortMonths,
            loading: false,
            dialog: false,
            deleteDialog: false,
            deleting: false,
            variety: this.getDefaultVariety(),
            headers: [
                {
                    value: 'name', text: this.$t('shared.name'), sortable: true, filterable: true,
                },
                {
                    value: 'exposure', text: this.$t('shared.exposure'), sortable: true, filterable: false,
                },
                {
                    value: 'watering', text: this.$t('shared.watering'), sortable: true, filterable: false,
                },
                {
                    value: 'timeline', text: '', sortable: false, filterable: false,
                },
                {
                    value: 'actions', text: '', sortable: false, filterable: false,
                },
            ],
        };
    },
    async created() {
        this.loading = true;
        await this.loadVarieties(this.genus);
        this.loading = false;
    },
    methods: {
        ...mapActions('genera', ['loadVarieties', 'deleteVariety']),

        getDefaultVariety() {
            return {
                genusId: this.genus.id,
                name: '',
                imageUrl: '',
                exposure: '',
                watering: '',
                seed: [],
                plant: [],
                harvest: [],
            };
        },
        onAdd() {
            this.variety = this.getDefaultVariety();
            this.dialog = true;
        },
        onEdit(variety) {
            this.variety = Vue.util.extend({}, variety);
            this.dialog = true;
        },
        onDelete(variety) {
            this.variety = variety;
            this.deleteDialog = true;
        },
        async onConfirmDelete() {
            this.deleting = true;
            await this.deleteVariety(this.variety);
            this.deleting = false;
            this.deleteDialog = false;
        },
    },
};
</script>
