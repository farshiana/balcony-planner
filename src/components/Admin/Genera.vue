<template>
    <v-container fluid>
        <v-data-table
            :headers="headers"
            :items="genera"
            :items-per-page="12"
            :loading="loadingGenera"
            show-group-by
            dense
            class="elevation-1"
            :item-class="() => 'clickable'"
            @click:row="onSelect"
        >
            <template  v-slot:top>
                <v-row fill-height class="mr-1">
                    <v-spacer />
                    <v-btn class="mt-1" text small color="accent" @click="onAdd">
                        {{ $t('admin.genera.addGenus') }}
                    </v-btn>
                </v-row>
            </template>
            <template v-slot:[`item.actions`]="{ item }">
                <v-icon
                    small
                    class="mr-2"
                    @click.stop="onEdit(item)"
                >
                    mdi-pencil
                </v-icon>
                <v-icon
                    small
                    @click.stop="onDelete(item)"
                >
                    mdi-delete
                </v-icon>
            </template>
        </v-data-table>
        <genus-form
            :visible="dialog"
            :genus="genus"
            @toggle="(visible) => { dialog = visible; }"
        />
        <v-dialog v-model="deleteDialog" persistent max-width="290">
            <v-card>
                <v-card-title class="headline">{{ $t('admin.genera.deleteGenus') }}</v-card-title>
                <v-card-text>{{ $t('admin.genera.deleteGenusConfirmation', { name: genus.name }) }}</v-card-text>
                <v-card-actions>
                    <v-spacer></v-spacer>
                    <v-btn text @click="deleteDialog = false;">{{ $t('shared.cancel') }}</v-btn>
                    <v-btn
                        color="error"
                        text
                        :loading="deleting"
                        @click="onConfirmDelete"
                    >
                        {{ $t('shared.confirm') }}
                    </v-btn>
                </v-card-actions>
            </v-card>
        </v-dialog>
        <router-view />
    </v-container>
</template>

<script>
import Vue from 'vue';
import { mapState, mapActions } from 'vuex';
import GenusForm from './GenusForm.vue';

export default {
    components: {
        GenusForm,
    },
    data() {
        return {
            dialog: false,
            deleteDialog: false,
            deleting: false,
            genus: this.getDefaultGenus(),
            headers: [
                { value: 'name', text: this.$t('shared.name'), groupable: false },
                { value: 'category', text: this.$t('admin.genera.category'), groupable: true },
                { value: 'actions', text: '', groupable: false },
            ],
        };
    },
    computed: {
        ...mapState('genera', ['genera', 'loadingGenera']),
    },
    created() {
        this.loadGenera();
    },
    methods: {
        ...mapActions('genera', ['loadGenera', 'deleteGenus']),

        getDefaultGenus: () => ({
            name: '',
            category: '',
        }),
        onAdd() {
            this.genus = this.getDefaultGenus();
            this.dialog = true;
        },
        onEdit(genus) {
            this.genus = Vue.util.extend({}, genus);
            this.dialog = true;
        },
        onDelete(genus) {
            this.genus = genus;
            this.deleteDialog = true;
        },
        async onConfirmDelete() {
            this.deleting = true;
            await this.deleteGenus(this.genus.id); // TODO: implement action
            this.deleting = false;
            this.deleteDialog = false;
        },
        onSelect(row) {
            this.$router.push({ name: 'genus', params: { genusId: row.id } });
        },
    },
};
</script>
