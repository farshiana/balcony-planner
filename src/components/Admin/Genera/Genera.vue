<template>
    <v-card class="ma-3" style="position: initial">
        <v-card-title>
            {{ $t('shared.genera') }}
            <v-spacer />
            <v-text-field
                v-model="search"
                append-icon="mdi-magnify"
                :label="$t('shared.searchPlaceholder')"
                single-line
                hide-details
            />
        </v-card-title>
        <v-data-table
            :headers="headers"
            :items="genera"
            :items-per-page="12"
            :loading="loadingGenera"
            show-group-by
            dense
            :search="search"
            class="genera"
            :item-class="() => 'clickable'"
            @click:row="onSelect"
        >
            <template  v-slot:top>
                <v-row fill-height class="mr-1">
                    <v-spacer />
                    <v-btn class="mt-1" text small color="accent" @click="onAdd">
                        {{ $t('admin.genera.addGenus') }}
                    </v-btn>
                </v-row>
            </template>
            <template v-slot:[`item.name`]="{ item }">
                <div class="text-capitalize-first">{{ item.name }}</div>
            </template>
            <template v-slot:[`item.actions`]="{ item }">
                <v-btn
                    icon
                    small
                    class="mr-2"
                    @click.stop="onEdit(item)"
                >
                    <v-icon small>mdi-pencil</v-icon>
                </v-btn>
                <v-btn
                    icon
                    small
                    @click.stop="onDelete(item)"
                >
                    <v-icon small>mdi-delete</v-icon>
                </v-btn>
            </template>
        </v-data-table>
        <genus-form
            :visible="dialog"
            :genus="genus"
            @toggle="(visible) => { dialog = visible; }"
        />
        <delete-dialog
            :visible.sync="deleteDialog"
            :title="$t('admin.genera.deleteGenus')"
            :name="genus.name"
            :deleting="deleting"
            @delete="onConfirmDelete"
        />
        <router-view />
    </v-card>
</template>

<script>
import Vue from 'vue';
import { mapState, mapActions } from 'vuex';
import DeleteDialog from '@/components/DeleteDialog.vue';
import GenusForm from './GenusForm.vue';

export default {
    components: {
        GenusForm,
        DeleteDialog,
    },
    data() {
        return {
            search: '',
            dialog: false,
            deleteDialog: false,
            deleting: false,
            genus: this.getDefaultGenus(),
            headers: [
                {
                    value: 'name',
                    text: this.$t('shared.genus'),
                    sortable: true,
                    groupable: false,
                    filterable: true,
                },
                {
                    value: 'category',
                    text: this.$t('admin.genera.category'),
                    sortable: true,
                    groupable: true,
                    filterable: true,
                },
                {
                    value: 'actions',
                    text: '',
                    sortable: false,
                    groupable: false,
                    filterable: false,
                },
            ],
        };
    },
    computed: {
        ...mapState('genera', ['genera', 'loadingGenera']),
    },
    created() {
        this.loadGenera();
    },
    methods: {
        ...mapActions('genera', ['loadGenera', 'deleteGenus']),

        getDefaultGenus: () => ({
            name: '',
            category: '',
        }),
        onAdd() {
            this.genus = this.getDefaultGenus();
            this.dialog = true;
        },
        onEdit(genus) {
            this.genus = Vue.util.extend({}, genus);
            this.dialog = true;
        },
        onDelete(genus) {
            this.genus = genus;
            this.deleteDialog = true;
        },
        async onConfirmDelete() {
            this.deleting = true;
            await this.deleteGenus(this.genus);
            this.deleting = false;
            this.deleteDialog = false;
        },
        onSelect(row) {
            this.$router.push({ name: 'genus', params: { genusId: row.id } });
        },
    },
};
</script>
