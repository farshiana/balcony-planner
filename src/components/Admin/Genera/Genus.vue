<template>
    <v-navigation-drawer
        v-model="drawer"
        absolute
        temporary
        right
        width="780px"
        @input="onInput"
    >
        <v-container>
            <div class="d-flex align-center justify-space-between">
                <span class="text-h5 text-capitalize-first">{{ genus.name }}</span>
                <v-btn
                    class="mr-1"
                    icon
                    @click="$router.push('genera')"
                >
                    <v-icon>mdi-close</v-icon>
                </v-btn>
            </div>
            <varieties :genus="genus" />
        </v-container>
    </v-navigation-drawer>
</template>

<script>
import { mapGetters } from 'vuex';
import Varieties from './Varieties.vue';

export default {
    components: {
        Varieties,
    },
    props: {
        genusId: {
            type: String,
            required: true,
        },
    },
    data() {
        return {
            drawer: true,
        };
    },
    computed: {
        ...mapGetters('genera', ['getGenusById']),

        genus() {
            return this.getGenusById(this.genusId) || { id: this.genusId, name: '', varieties: [] };
        },
    },
    methods: {
        onInput(value) {
            if (!value) {
                this.$router.push({ name: 'genera' });
            }
        },
    },
};
</script>
