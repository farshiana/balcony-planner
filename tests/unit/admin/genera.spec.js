import { fireEvent } from '@testing-library/vue';
import Genera from '@/components/Admin/Genera/Genera.vue';
import { CATEGORY_FRUITS } from '@/constants';
import render from '../../render';

describe('Genera.vue', () => {
    beforeEach(() => {
        fetch.mockResponseOnce(JSON.stringify([
            { name: 'genus1', category: CATEGORY_FRUITS },
            { name: 'genus2', category: CATEGORY_FRUITS },
        ]));
    });

    it('lists genera', async () => {
        const { container, findByText } = render(Genera);
        await findByText('genus1');
        await findByText('genus2');
        expect(container).toMatchSnapshot();
        expect(fetch).toHaveBeenCalledTimes(1);
    });

    it('creates genus', async () => {
        fetch.mockResponseOnce(JSON.stringify({ id: 1 }));

        const { getByText, getByLabelText, getByTestId } = render(Genera);
        await fireEvent.click(getByText('Add genus'));
        await fireEvent.click(getByLabelText('Category'));
        await fireEvent.click(getByText('Fruits'));
        await fireEvent.update(getByLabelText('Name'), 'toto');
        await fireEvent.click(getByTestId('save-genus'));
        expect(fetch).toHaveBeenCalledTimes(3);
    });

    it('filters genera', async () => {
        const { getByLabelText, findByText, queryByText } = render(Genera);
        await fireEvent.update(getByLabelText('Search'), 'genus2');
        expect(queryByText('genus1')).not.toBeInTheDocument();
        await findByText('genus2');
    });
});
