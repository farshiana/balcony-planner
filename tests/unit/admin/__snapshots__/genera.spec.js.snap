// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Genera.vue lists genera 1`] = `
<div>
  <div
    class="ma-3 v-card v-sheet theme--light"
    style="position: initial;"
  >
    <div
      class="v-card__title"
    >
      
        Genera
        
      <div
        class="spacer"
      />
       
      <div
        class="v-input v-input--hide-details theme--light v-text-field v-text-field--single-line"
      >
        <div
          class="v-input__control"
        >
          <div
            class="v-input__slot"
          >
            <div
              class="v-text-field__slot"
            >
              <label
                class="v-label theme--light"
                for="input-9"
                style="left: 0px; position: absolute;"
              >
                Search
              </label>
              <input
                id="input-9"
                type="text"
              />
            </div>
            <div
              class="v-input__append-inner"
            >
              <div
                class="v-input__icon v-input__icon--append"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-magnify theme--light"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
     
    <div
      class="v-data-table genera v-data-table--dense theme--light"
    >
      <div
        class="row mr-1"
        fill-height=""
      >
        <div
          class="spacer"
        />
         
        <button
          class="mt-1 v-btn v-btn--flat v-btn--text theme--light v-size--small accent--text"
          type="button"
        >
          <span
            class="v-btn__content"
          >
            
                    Add genus
                
          </span>
        </button>
      </div>
      <div
        class="v-data-table__wrapper"
      >
        <table>
          <colgroup>
            <col
              class=""
            />
            <col
              class=""
            />
            <col
              class=""
            />
          </colgroup>
          <thead
            class="v-data-table-header"
          >
            <tr>
              <th
                aria-label="Genus: Not sorted. Activate to sort ascending."
                aria-sort="none"
                class="text-start sortable"
                role="columnheader"
                scope="col"
              >
                <span>
                  Genus
                </span>
                <i
                  aria-hidden="true"
                  class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light"
                  style="font-size: 18px;"
                />
              </th>
              <th
                aria-label="Category: Not sorted. Activate to sort ascending."
                aria-sort="none"
                class="text-start sortable"
                role="columnheader"
                scope="col"
              >
                <span>
                  Category
                </span>
                <i
                  aria-hidden="true"
                  class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light"
                  style="font-size: 18px;"
                />
                <span>
                  group
                </span>
              </th>
              <th
                aria-label=""
                class="text-start"
                role="columnheader"
                scope="col"
              >
                <span />
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              class="clickable"
            >
              <td
                class="text-start"
              >
                <div
                  class="text-capitalize-first"
                >
                  genus1
                </div>
              </td>
              <td
                class="text-start"
              >
                fruits
              </td>
              <td
                class="text-start"
              >
                <button
                  class="mr-2 v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small"
                  type="button"
                >
                  <span
                    class="v-btn__content"
                  >
                    <i
                      aria-hidden="true"
                      class="v-icon notranslate mdi mdi-pencil theme--light"
                      style="font-size: 16px;"
                    />
                  </span>
                </button>
                 
                <button
                  class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small"
                  type="button"
                >
                  <span
                    class="v-btn__content"
                  >
                    <i
                      aria-hidden="true"
                      class="v-icon notranslate mdi mdi-delete theme--light"
                      style="font-size: 16px;"
                    />
                  </span>
                </button>
              </td>
            </tr>
            <tr
              class="clickable"
            >
              <td
                class="text-start"
              >
                <div
                  class="text-capitalize-first"
                >
                  genus2
                </div>
              </td>
              <td
                class="text-start"
              >
                fruits
              </td>
              <td
                class="text-start"
              >
                <button
                  class="mr-2 v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small"
                  type="button"
                >
                  <span
                    class="v-btn__content"
                  >
                    <i
                      aria-hidden="true"
                      class="v-icon notranslate mdi mdi-pencil theme--light"
                      style="font-size: 16px;"
                    />
                  </span>
                </button>
                 
                <button
                  class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small"
                  type="button"
                >
                  <span
                    class="v-btn__content"
                  >
                    <i
                      aria-hidden="true"
                      class="v-icon notranslate mdi mdi-delete theme--light"
                      style="font-size: 16px;"
                    />
                  </span>
                </button>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        class="v-data-footer"
      >
        <div
          class="v-data-footer__select"
        >
          Rows per page:
          <div
            class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select"
          >
            <div
              class="v-input__control"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-owns="list-21"
                class="v-input__slot"
                role="button"
              >
                <div
                  class="v-select__slot"
                >
                  <div
                    class="v-select__selections"
                  >
                    <div
                      class="v-select__selection v-select__selection--comma"
                    >
                      5
                    </div>
                    <input
                      aria-label="$vuetify.dataTable.itemsPerPageText"
                      aria-readonly="false"
                      autocomplete="off"
                      id="input-21"
                      readonly="readonly"
                      type="text"
                    />
                  </div>
                  <div
                    class="v-input__append-inner"
                  >
                    <div
                      class="v-input__icon v-input__icon--append"
                    >
                      <i
                        aria-hidden="true"
                        class="v-icon notranslate mdi mdi-menu-down theme--light"
                      />
                    </div>
                  </div>
                  <input
                    type="hidden"
                    value="[object Object]"
                  />
                </div>
                <div
                  class="v-menu"
                >
                  <!---->
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="v-data-footer__pagination"
        >
          1-2 of 2
        </div>
        <div
          class="v-data-footer__icons-before"
        >
          <button
            aria-label="Previous page"
            class="v-btn v-btn--disabled v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default"
            disabled="disabled"
            type="button"
          >
            <span
              class="v-btn__content"
            >
              <i
                aria-hidden="true"
                class="v-icon notranslate mdi mdi-chevron-left theme--light"
              />
            </span>
          </button>
        </div>
        <div
          class="v-data-footer__icons-after"
        >
          <button
            aria-label="Next page"
            class="v-btn v-btn--disabled v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default"
            disabled="disabled"
            type="button"
          >
            <span
              class="v-btn__content"
            >
              <i
                aria-hidden="true"
                class="v-icon notranslate mdi mdi-chevron-right theme--light"
              />
            </span>
          </button>
        </div>
      </div>
    </div>
     
    <div
      class="v-dialog__container"
      role="dialog"
    >
      <!---->
    </div>
     
    <div
      class="v-dialog__container"
      role="dialog"
    >
      <!---->
    </div>
     
    <!---->
  </div>
</div>
`;
